roqoqo/tests/integration/circuitdag.rs

```diff
@@ -0,0 +1,472 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the
+// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use roqoqo::operations::*;
+use roqoqo::{Circuit, CircuitDag};
+
+use test_case::test_case;
+
+/// Test adding an operation that doesn't involve qubits.
+///
+#[test_case(
+    Operation::from(PragmaSetNumberOfMeasurements::new(3, "ro".to_string())),
+    Operation::from(PragmaSetNumberOfMeasurements::new(5, "ti".to_string()))
+)]
+#[test_case(
+    Operation::from(PragmaSetNumberOfMeasurements::new(4, "ri".to_string())),
+    Operation::from(PragmaSetNumberOfMeasurements::new(1, "to".to_string()))
+)]
+fn add_operation_no_involved_qubits(operation1: Operation, operation2: Operation) {
+    let mut dag: CircuitDag = CircuitDag::new();
+
+    let back1 = dag.add_to_back(operation1.clone());
+
+    assert!(operation1.involved_qubits() == InvolvedQubits::None);
+    assert_eq!(dag.commuting_operations().get(0), back1.as_ref());
+
+    dag.add_to_front(Operation::from(PauliY::new(0)));
+    dag.add_to_back(Operation::from(CNOT::new(0, 1)));
+
+    let front1 = dag.add_to_front(operation2.clone());
+
+    assert_eq!(dag.commuting_operations().get(1), front1.as_ref());
+}
+
+/// Test graph node existance after adding an operation that involves qubits.
+///
+#[test_case(Operation::from(PauliX::new(0)))]
+#[test_case(Operation::from(PauliY::new(1)))]
+#[test_case(Operation::from(ControlledPauliZ::new(0, 1)))]
+fn check_node_existance(operation: Operation) {
+    let mut dag: CircuitDag = CircuitDag::new();
+
+    dag.add_to_back(operation.clone());
+
+    assert!(dag.graph().node_count() == 1);
+
+    dag.add_to_front(operation.clone());
+    dag.add_to_back(Operation::from(CNOT::new(0, 1)));
+
+    assert!(dag.graph().node_count() == 3);
+}
+
+#[test_case(Operation::from(PauliX::new(0)), Operation::from(PauliY::new(0)))]
+#[test_case(Operation::from(PauliZ::new(0)), Operation::from(CNOT::new(0, 1)))]
+fn check_node_count(operation1: Operation, operation2: Operation) {
+    let mut dag: CircuitDag = CircuitDag::new();
+
+    dag.add_to_back(operation1.clone());
+    dag.add_to_front(operation2.clone());
+
+    assert!(dag.graph().node_count() == 2);
+
+    dag.add_to_back(operation1.clone());
+
+    assert!(dag.graph().node_count() == 3);
+}
+
+#[test_case(Operation::from(PauliX::new(0)), Operation::from(PauliY::new(0)))]
+fn check_edge(operation1: Operation, operation2: Operation) {
+    let mut dag: CircuitDag = CircuitDag::new();
+
+    let ind1 = dag.add_to_back(operation1.clone());
+    let ind2 = dag.add_to_back(operation2.clone());
+
+    assert!(dag
+        .graph()
+        .contains_edge(ind1.unwrap().into(), ind2.unwrap().into()));
+}
+
```

Comment:
Could you also add a test for edges if we have a all-qubit-operation in the circuit as this is a special case?